### ISSUE TITLE
A question about MAN

### ISSUE THREAD

**Issue by KID0031** [keqinw/MAN#1]
Hi, I read the MAN paper and tried to run it in the Mujoco environment like SDQN did. However, the performance was not very good. Since MAN and SDQN share similar ideas, their performance in Mujoco should at least be comparable. Is there any possible reason for this or maybe I made some mistakes? By the way, I used the code in the reacher environment and modified it to Mujoco.

Thanks a lot :)

**Comment by keqinw** [comment]
Hi, thanks for your reading, and thanks for you work to adapt MAN into Mujoco! 

To you question, here are some suggestions that might help you to locate the problem:
1. Please pay attention that the processes in MAN and SDQN are autoregessive, which means the input of next Q-net is the combine of previous state and the previously predicted 1-D actions.  
2. The idea of MAN and SDQN are diffrent actually: in SDQN, they used a whole model (like a LSTM) to make the decision, while in MAN, we used several different sub-action networks to make the decision. Therefore, it is harder to train MAN since the networks are seperated to each other. (But it also has some advantages, like what I mentioned in #4)
3. I saw that you mentioned Mujoco, therefore I guess that you might try to use MAN or SDQN to solve some continuous control by discretizing the action space. If so, however, I have to say that I failed in this case with MAN previously.  The potential explanation is that the action space is still too large to the MAN's capability. One solution is keeping fine discretizing in some important action dimensions, and get rough discretizing in other unimportants. 
4. As a conclusion, I think it makes sense that you have two different results with large gap between MAN and SDQN. It highly depend on the training task. The advantage of MAN is that you do not need to fix the action number in each dimention, since you can utilize sub-actions networks with different output layer. It's trade-off. 

Hope my answer could help you have a deeper understanding of MAN and polish your project. Thanks for reading my paper!

**Comment by KID0031** [comment]
Thank you for your reply! I also have a question about the code in `train_man.py` in the Reacher environment. It is written that `y = torch.cat([y,sample_reward + self.gamma * torch.max(self.agent.target_nets[j](u_states_next).detach())])`. Is it should be `torch.max(self.agent.target_nets[j](u_states_next[-1]).detach())])` because it uses the maximum value of Q in the transition instead of the whole batch?

**Comment by KID0031** [comment]
Thank you for your reply! I also have a question about the code in train_man.py in the Reacher environment. It is written that `y = torch.cat([y,sample_reward + self.gamma * torch.max(self.agent.target_nets[j](u_states_next).detach())])`. Is it should be ` torch.max(self.agent.target_nets[j](u_states_next[-1]).detach())])` because it uses the maximum value of Q in the transition instead of the whole batch?

**Comment by keqinw** [comment]
Yes you are right, that's a good catch! Much appreciated for pointing out this error. And your solution works, or more concisely, just use `u_state_next` in line 128 and delete the line 114. You can check the new version that I uploaded. 
As for the result, for me, I just have run several experiments and there was no big difference between the modified and the un-modified version. If you have some new result, please let me know. 

Thanks again for your careful reading!

**Comment by KID0031** [comment]
Thanks for the prompt reply! As you said, the difference between the two is not significant. 

I have another question about the code, and I may not have fully understood it(sorry for having so many questions). Here `u_state_next = torch.cat([sample_next_state,sample_action[:,:j]],dim = 1)` the `s_{t+1}` is concatenated with sampled previous timestep actions `a_t[:j]`, so why is it not using `a_{t+1}[:j]`, because in equation (3) in the paper, `u_{t+1}^{k+1}` should be `(s_{t+1}, a_{t+1}[:k])` instead of `a_{t}[:k]`.

**Comment by keqinw** [comment]
Questions are welcomed!

The answer to this question is **keeping the** `a_t[:j]`. You may have this trouble since I have a typro in equation (3), while the correct one should be like:
```math
u^{k+1}_{t+1} = (s_{t+1}, a^{1:k}_t)
``` 
Actually, I think here I need to change $u$ to another new character such as $v$, in order to distinguish it with the $u^k_t$ in the same equation.

Here is the explanation: 
1. First of all, it's weird that using `a_{t+1}[:j]` here right? Because that nobody could know what will happen in the future (at timestep $t+1$ here).
2. The reason why we concatenated $s_{t+1}$ with $a^k_{t}$ is that, you can regard every Q-net ($Q^k$) as a function of the state ($s$) and the action at the corresponding dimention $k$ ($a^{k}$). Therefore, the actions at other dimension, should not be the variables in this function. Instead, they should be constants. That's why we use `a_t[:j]` here, keeping the same within $Q^k$ and $Q^{k+1}$. Then when you do back propagation, you would pay attention on $s$ and $a^{k}$, and ignore the other actions at other dimension (since the derivative of constant is 0). Does it make sense?

Thansk for your questions again, and I will update the paper later on.
